src/app.jsx

```diff
@@ -44,23 +50,101 @@ export default class App extends React.Component {
     this.handleFileSelect = this.handleFileSelect.bind(this);
     this.onSelectData = this.onSelectData.bind(this);
     this.onSelectColorPack = this.onSelectColorPack.bind(this);
+    this.toggleUserData = this.toggleUserData.bind(this);
+    this.editHeader = this.editHeader.bind(this);
+    this.saveEditedHeaders = this.saveEditedHeaders.bind(this);
+    this.addHeaders = this.addHeaders.bind(this);
+    this.showHeaderDialog = this.showHeaderDialog.bind(this);
+    this.cancelEditDialog = this.cancelEditDialog.bind(this);
   }
 
   handleFileSelect(evt) {
     const file = evt.target.files[0];
 
     Papa.parse(file, {
       complete: (results) => {
-        this.setState({data: results.data});
+        this.setState({data: results.data, headerRow: results.data[0]});
       },
     });
   }
 
+  toggleUserData(evt) {
+    if (evt.target.value === 'true') {
+      this.setState({useSampleData: true, data: data.smallData});
+    } else {
+      this.setState({useSampleData: false, data: [[]]});
+    }
+  }
+
+  showHeaderDialog(e) {
+    const editType = e.target.value;
+
+    if (editType === 'editHeaders') {
+      this.setState({showEditHeaderForm: true, showAddHeaderForm: false,
+        headerRow: this.state.data[0]});
+    } else if (editType === 'addHeaders') {
+      const blankHeaderRow = [];
+
+      this.state.data[0].forEach(header => {
+        blankHeaderRow.push(null);
+      });
+      this.setState({showAddHeaderForm: true, showEditHeaderForm: false,
+        headerRow: blankHeaderRow});
+    }
+  }
+
+  cancelEditDialog() {
+    this.setState({showEditHeaderForm: false, showAddHeaderForm: false,
+      headerRow: this.state.data[0], validationMessage: null});
+  }
+
+  saveEditedHeaders() {
+    const newDataSet = this.state.data.slice();
+
+    newDataSet[0] = this.state.headerRow;
+    this.setState({data: newDataSet, showEditHeaderForm: false});
+  }
+
+  addHeaders() {
+    const newDataSet = this.state.data.slice();
+
+    let validHeaders = true;
```

Comment:
You could use the use the reduce patten instead to declare this variable. See below. I'll leave it to you if you want to change it or not though!

```
    const validHeaders = this.state.headerRow.reduce((acc, header) => {
      if (header === null || header === '') {
        acc = false;
      }
      return acc;
    }, true);
```